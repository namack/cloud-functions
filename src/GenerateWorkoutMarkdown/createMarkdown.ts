import moment from 'moment';
import pluralize from 'pluralize';
import { addCommas } from '../utils';
import { StravaDetailedActivity } from './stravaTypes';

const createMarkdown = (activity: StravaDetailedActivity): string => {
  const heartrateSection =
    activity.has_heartrate &&
    `
_Average Heart Rate_

### ${Math.floor(activity.average_heartrate)} BPM

_Max Heart Rate_

### ${Math.floor(activity.max_heartrate)} BPM
`.trim();

  const distance = activity.distance / 1609.34;
  const hours = parseInt(
    moment.utc(activity.moving_time * 1000).format('h'),
    10
  );
  const minutes = parseInt(
    moment.utc(activity.moving_time * 1000).format('m'),
    10
  );

  const elapsedTime = () => {
    const hourString = `${hours} ${pluralize('hour', hours)}`;
    const minuteString = `${minutes} ${pluralize('minute', minutes)}`;
    if (hours > 0 && minutes > 0) {
      return `${hourString}, ${minuteString}`;
    } else if (hours === 0 && minutes > 0) {
      return minuteString;
    } else if (hours > 0 && minutes === 0) {
      return hourString;
    }
  };

  const calories = addCommas(JSON.stringify(Math.floor(activity.calories)));

  const averageSpeed = (): string | undefined => {
    const speed = activity.average_speed * 2.237;
    if (activity.type === 'VirtualRide' || activity.type === 'Ride') {
      return `${speed.toFixed(2)} MPH`;
    } else if (activity.type === 'Run' || activity.type === 'VirtualRun') {
      const pace = 60 / speed;
      return `${pace.toFixed(2)} / Mile`;
    }
  };

  const paceSection =
    averageSpeed &&
    `
_Average Pace_

### ${averageSpeed()}
`.trim();

  const elevationSection =
    (activity.type === 'Run' ||
      activity.type === 'Ride' ||
      activity.type === 'VirtualRide' ||
      activity.type === 'VirtualRun') &&
    `
_Elevation Gain_

### ${addCommas(
      JSON.stringify(Math.floor(activity.total_elevation_gain * 3.281))
    )} feet
`.trim();

  return `
## [${activity.name}](https://www.strava.com/activities/${activity.id})

_Distance_

### ${distance.toFixed(2)} miles

_Total Time_

### ${elapsedTime()}

${paceSection}

${elevationSection}

${heartrateSection}

_Calories_

### ${calories} Calories
`.trim();
};

export { createMarkdown };
