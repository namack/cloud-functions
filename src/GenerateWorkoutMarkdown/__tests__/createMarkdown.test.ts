import { createMarkdown } from '../createMarkdown';
import bike from './__fixtures__/bike.json';
import run from './__fixtures__/run.json';

test('markdown post generation', () => {
  expect(createMarkdown(bike)).toMatchSnapshot();

  expect(createMarkdown(run)).toMatchSnapshot();
});
