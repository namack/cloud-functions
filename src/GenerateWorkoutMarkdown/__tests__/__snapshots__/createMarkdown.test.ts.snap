// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown post generation 1`] = `
"## [Build Me Up - Novanta](https://www.strava.com/activities/2214831402)

_Distance_

### 19.68 miles

_Total Time_

### 1 hour

_Average Pace_

### 19.61 MPH

_Elevation Gain_

### 715 feet

_Average Heart Rate_

### 154 BPM

_Max Heart Rate_

### 180 BPM

_Calories_

### 604 Calories"
`;

exports[`markdown post generation 2`] = `
"## [Afternoon Run](https://www.strava.com/activities/2221133947)

_Distance_

### 15.26 miles

_Total Time_

### 2 hours, 6 minutes

_Average Pace_

### 8.27 / Mile

_Elevation Gain_

### 281 feet

_Average Heart Rate_

### 150 BPM

_Max Heart Rate_

### 179 BPM

_Calories_

### 2,404 Calories"
`;
